message_ix_models/tools/inter_pipe/generate_inter_pipe.py

```diff
@@ -0,0 +1,796 @@
+
+
+# Small util function
+def get_template(scen_base, par_name, tech_mother_pipe):
+    template = pd.DataFrame()
+    template = (
+        scen_base.par(par_name, filters={"technology": tech_mother_pipe})
+        .head()
+        .iloc[0]
+        .to_frame()
+        .T
+    )
+    if template.empty:
+        log.warning(
+            f"Technology {tech_mother_pipe} does not have {par_name} in {scen_base}."
+        )
+    return template
+
+
+# Small util function
+def copy_template_columns(df, template, exclude_cols=["node_loc", "technology"]):
+    for col in template.columns:
+        if col not in exclude_cols:
+            df[col] = template[col].iloc[0]
+
+
+# Small util function
+def broadcast_yv_ya(df: pd.DataFrame, ya_list: list[int]) -> pd.DataFrame:
+    """
+    Broadcast years to create vintage-activity year pairs.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        Input parameter DataFrame
+    ya_list : list[int]
+        List of activity years to consider
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame with expanded rows for each vintage-activity year pair
+    """
+    all_new_rows = []
+    # Process each row in the original DataFrame
+    for _, row in df.iterrows():
+        # For each activity year
+        for ya in ya_list:
+            # Get all vintage years that are <= activity year
+            yv_list = [yv for yv in ya_list if yv <= ya]
+
+            # Create new rows for each vintage year
+            for yv in yv_list:
+                new_row = row.copy()
+                new_row["year_act"] = int(ya)
+                new_row["year_vtg"] = int(yv)
+                all_new_rows.append(new_row)
+    # Combine original DataFrame with new rows
+    result_df = pd.concat([df, pd.DataFrame(all_new_rows)], ignore_index=True)
+    result_df = result_df[result_df["year_vtg"] != "broadcast"]
+    return result_df
+
+
+# Small util function
+def broadcast_yv(df: pd.DataFrame, ya_list: list[int]) -> pd.DataFrame:
+    """Broadcast vintage years."""
+    all_new_rows = []
+    for _, row in df.iterrows():
+        for yv in ya_list:
+            new_row = row.copy()
+            new_row["year_vtg"] = int(yv)
+            all_new_rows.append(new_row)
+    result_df = pd.concat([df, pd.DataFrame(all_new_rows)], ignore_index=True)
+    result_df = result_df[result_df["year_vtg"] != "broadcast"]
+    return result_df
+
+
+# Small util function
+def broadcast_yl(df: pd.DataFrame, ya_list: list[int]) -> pd.DataFrame:
+    """Broadcast relation years."""
+    all_new_rows = []
+    for _, row in df.iterrows():
+        for yv in ya_list:
+            new_row = row.copy()
+            new_row["year_rel"] = int(yv)
+            all_new_rows.append(new_row)
+    result_df = pd.concat([df, pd.DataFrame(all_new_rows)], ignore_index=True)
+    result_df = result_df[result_df["year_rel"] != "broadcast"]
+    return result_df
+
+
+# Main function to generate bare sheets
+def inter_pipe_bare(
```

Comment:
If we look at the "fully qualified name" (FQN) here, we see:

`message_ix_models.tools.inter_pipe.generate_inter_pipe.inter_pipe_bare()`

In this, the string "inter_pipe" occurs three times. Except in rare cases, a FQN should probably not contain repetition.

We also see that .inter_pipe has an empty \_\_init__.py file and only one submodule.

I would suggest to rearrange such this function has a FQN like:

`message_ix_models.tools.inter_pipe.make_template()`